src/box/txn_limbo.c

```diff
@@ -107,12 +107,12 @@ txn_limbo_append(struct txn_limbo *limbo, uint32_t id, struct txn *txn)
 		diag_set(ClientError, ER_SYNC_ROLLBACK);
 		return NULL;
 	}
-	if (id == 0)
-		id = instance_id;
+	/* id 0 is only possible for local transactions. */
+	assert(id != 0 || !txn_has_flag(txn, TXN_WAIT_ACK));
 	if  (limbo->owner_id == REPLICA_ID_NIL) {
 		diag_set(ClientError, ER_SYNC_QUEUE_UNCLAIMED);
 		return NULL;
-	} else if (limbo->owner_id != id) {
+	} else if (limbo->owner_id != id && id != REPLICA_ID_NIL) {
```

Comment:
Wouldn't this patch be simpler?
```Diff
diff --git a/src/box/txn.h b/src/box/txn.h
index 9ac7441f3..b7252ed70 100644
--- a/src/box/txn.h
+++ b/src/box/txn.h
@@ -845,6 +845,13 @@ txn_commit_ro_stmt(struct txn *txn, struct txn_ro_savepoint *svp)
 bool
 txn_is_distributed(struct txn *txn);
 
+static inline bool
+txn_is_fully_local(const struct txn *txn)
+{
+       return txn->n_new_rows == txn->n_local_rows &&
+              txn->n_applier_rows == 0;
+}
+
 /**
  * End a statement. In autocommit mode, end
  * the current transaction as well.
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index acdde6695..2d838413f 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -112,7 +112,7 @@ txn_limbo_append(struct txn_limbo *limbo, uint32_t id, struct txn *txn)
        if  (limbo->owner_id == REPLICA_ID_NIL) {
                diag_set(ClientError, ER_SYNC_QUEUE_UNCLAIMED);
                return NULL;
-       } else if (limbo->owner_id != id) {
+       } else if (limbo->owner_id != id && !txn_is_fully_local(txn)) {
                if (txn_limbo_is_empty(limbo)) {
                        diag_set(ClientError, ER_SYNC_QUEUE_FOREIGN,
                                 limbo->owner_id);
```
It allows to avoid origin id manipulations in txn.c.

Reply:
> Wouldn't this patch be simpler?

It would, thanks! Applied.